sapicore/engine/synapse/__init__.py

```diff
@@ -0,0 +1,330 @@
+
+    delay_ms: float or Tensor
+        Generic transmission delay, managed by the synapse object using queues.
+        Transmission delays are useful for some temporal coding schemes.
+
+    simple_delays: bool
+        Whether transmission delay values from one presynaptic element are the same for all postsynaptic targets.
+        Provides a speed-control tradeoff (one matrix multiplication vs. N vector multiplications).
+
+    weight_init_method: Callable
+        Weight initialization method from `torch.nn.init`.
+
+    connections: Tensor
+        2D binary integer mask matrix indicating which source unit -> destination unit connections are enabled.
+        Loggable attribute.
+
+    weights: Tensor
+        2D float matrix storing a weight for each pair of source and destination ensemble units.
+        Loggable attribute.
+
+    output: Tensor
+        1D tensor storing the result of W*x matrix multiplication in the current simulation step.
+        Loggable attribute.
+
+    """
+
+    _config_props_: tuple[str] = ("weight_max", "weight_min", "delay_ms", "simple_delays")
+    _loggable_props_: tuple[str] = ("weights", "connections", "output")
+
+    # declare loggable instance attributes registered as torch buffers.
+    connections: tensor
+    weights: tensor
+    output: tensor
+
+    def __init__(
+        self,
+        src_ensemble: [Neuron] = None,
+        dst_ensemble: [Neuron] = None,
+        weight_max: float = 1000.0,
+        weight_min: float = -1000.0,
+        delay_ms: float = 0.0,
+        simple_delays: bool = True,
+        weight_init_method: Callable = xavier_uniform_,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        # model-related common instance attributes.
+        self.src_ensemble = src_ensemble
+        self.dst_ensemble = dst_ensemble
+
+        try:
+            self.matrix_shape = (dst_ensemble.num_units, src_ensemble.num_units)
+
+        except AttributeError:
+            logging.info("Attempted to initialize synapse object with no source and destination ensembles.")
+            self.matrix_shape = (1, 1)
+
+        self.num_units = torch.prod(torch.tensor(self.matrix_shape), 0).item()
+
+        self.weight_max = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + weight_max
+        self.weight_min = torch.zeros(self.matrix_shape, dtype=torch.float, device=self.device) + weight_min
+        self.weight_init_method = weight_init_method  # xavier_uniform by default.
+
+        # whether transmission delays from one presynaptic element are shared among postsynaptic targets.
+        self.simple_delays = simple_delays
+
+        if self.simple_delays:
+            # initialize default transmission delay values as a tensor of size `src_ensemble.num_units`.
+            self.delay_ms = torch.zeros(self.matrix_shape[1], dtype=torch.float, device=self.device) + delay_ms
```

Comment:
Overall, queues is an easy way to get delays working, but long term you'll probably need something more sophisticated because queues can be ineffecient, especially for the non-simple delays. Something like a ring matrix that you can copy in temp data, maybe.

But, for simple delays, where everything is delayed by the same value, I'm not sure why we need individual queues for each src neuron? Could you not just have one queue on n items, where each items is an array?

Reply:
Simple delays are not quite that--they allow for variable delays mapped many-to-one (i.e., all neurons in the destination layer have the same "view" of the delay of each source unit, but the latter can vary). This means a single matrix multiplication and a significant decrease in runtime while still allowing for some degree of delay heterogeneity.

That said, we do need a `no_delays` setting--if delays are of no interest (fixed), there's no reason to maintain queues. I'll do that in this PR and open an issue for optimizing queues.

Reply:
Hmm I may have misunderstood how this works.